File: examples/ocaml/linear/band.ml

open Sundials
open Bigarray

module M = Matrix.Band
let printf = Format.printf
let fprintf = Format.fprintf

let print_mat out m =
  let {M.n; M.mu; M.ml; M.smu} = M.dims m in
  fprintf out "@[<v>";
  for i = 0 to n - 1 do
    fprintf out "@[<h>";
    for j = 0 to n - 1 do
      if (i > j + ml) || (j > i + mu)
      then fprintf out "       --     "
      else fprintf out "@ % e" (M.get m i j)
    done;
    fprintf out "@]@ "
  done;
  fprintf out "@]"

let print_factored_mat out m =
  let {M.n; M.mu; M.ml; M.smu} = M.dims m in
  fprintf out "@[<v>";
  for i = 0 to n - 1 do
    fprintf out "@[<h>";
    for j = 0 to n - 1 do
      if (j > i + mu) && (j <= i + smu)
      then fprintf out "@ (% e)" (M.get m i j)
      else if (i > j + ml) || (j > i + mu)
      then fprintf out "        --      "
      else fprintf out "@  % e " (M.get m i j)
    done;
    fprintf out "@]@ "
  done;
  fprintf out "@]"

let print_vec out m =
  let nr = Array1.dim m in
  fprintf out "@[<h>";
  for i = 0 to nr - 1 do
    fprintf out "@ % e" m.{i}
  done;
  fprintf out "@]"

let print_p out p =
  let nc = Array1.dim p in
  fprintf out "@[<h>";
  for i = 0 to nc - 1 do
    fprintf out "@ % d" p.{i}
  done;
  fprintf out "@]"

let n =5
and mu = 1
and ml = 1;;
let smu = min (n - 1) (mu + ml);;

let main () =
  let a = M.create {M.n; M.mu; M.ml; M.smu} in
  let zero = M.make {M.n; M.mu; M.ml; M.smu} 0.0 in

  M.set a 0 0 ( 1.0);
  M.set a 0 1 ( 2.0);

  M.set a 1 0 ( 2.0);
  M.set a 1 1 ( 2.0);
  M.set a 1 2 ( 3.0);

  M.set a 2 1 ( 3.0);
  M.set a 2 2 ( 3.0);
  M.set a 2 3 ( 4.0);

  M.set a 3 2 ( 4.0);
  M.set a 3 3 ( 4.0);
  M.set a 3 4 ( 5.0);

  M.set a 4 3 ( 5.0);
  M.set a 4 4 ( 5.0);

  printf "initially: a=@\n%a@\n" print_mat a;

  (try
    let x = RealArray.of_array [| 1.0; 2.0; 3.0; 4.0; 5.0 |] in
    let y = RealArray.create n in
    M.matvec a x y;
    printf "matvec: y=@\n%a@\n\n" print_vec y
  with Config.NotImplementedBySundialsVersion -> ());

  let b = M.create {M.n; M.mu; M.ml; M.smu} in
  M.blit ~src:a ~dst:b;

  M.scale_add 2.0 b zero;
  printf "scale copy x2: b=@\n%a@\n" print_mat b;

  M.scale_addi 1.0 b;
  printf "add identity: b=@\n%a@\n" print_mat b;

  let a_ra2 = Sundials.RealArray2.wrap (M.unwrap a) in

  let p = LintArray.create 5 in
  Array1.fill p 0;
  Matrix.ArrayBand.gbtrf (a_ra2, (smu, mu, ml)) p;
  printf "getrf: a=@\n%a@\n" print_factored_mat a;
  printf "       p=@\n%a@\n@\n" print_p p;

  let s = RealArray.create n in
  s.{0} <-  5.0;
  s.{1} <- 15.0;
  s.{2} <- 31.0;
  s.{3} <- 53.0;
  s.{4} <- 45.0;
  Matrix.ArrayBand.gbtrs (a_ra2, (smu, mu, ml)) p s;
  printf "getrs: s=@\n%a@\n" print_vec s;;

main ();;
Gc.compact ();;

File: examples/ocaml/linear/band.c
#include <stdio.h>
#include <stdlib.h>
#include <sundials/sundials_types.h>
#include <sundials/sundials_direct.h>
#include <sundials/sundials_band.h>
#include "../../../src/config.h"
#include "../../../src/sundials/sundials_ml.h"

#define SIZE 5

#define ML 1
#define MU 1
#define SMU 2	/* min(SIZE -1, MU + ML) */

void print_mat(DlsMat m) {
    int i, j;

    for (i=0; i < m->M; ++i) {
	for (j=0; j < m->N; ++j) {
	    if ((i > j + m->ml) || (j > i + m->mu)) {
		printf("       --     ");
	    } else {
		printf(" % e", BAND_ELEM(m, i, j));
	    }
	}
	printf("\n");
    }
}

void print_factored_mat(DlsMat m) {
    int i, j;

    for (i=0; i < m->M; ++i) {
	for (j=0; j < m->N; ++j) {
	    if ((j > i + m->mu) && (j <= i + m->s_mu)) {
		printf(" (% e)", BAND_ELEM(m, i, j));
	    } else if ((i > j + m->ml) || (j > i + m->s_mu)) {
		printf("        --      ");
	    } else {
		printf("  % e ", BAND_ELEM(m, i, j));
	    }
	}
	printf("\n");
    }
}

void print_vec(sunrealtype* m, sundials_ml_index nr)
{
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % e", m[i]);
    }
    printf("\n");
}

void print_pivots(sundials_ml_index* m, sundials_ml_index nr)
{
    int i;

    for (i=0; i < nr; ++i) {
	printf(" % lld", (long long)m[i]);
    }
    printf("\n");
}

int main(int argc, char** argv)
{
#if 600 <= SUNDIALS_LIB_VERSION
    DlsMat a = SUNDlsMat_NewBandMat(SIZE, MU, ML, SMU);
    DlsMat b = SUNDlsMat_NewBandMat(SIZE, MU, ML, SMU);
#else
    DlsMat a = NewBandMat(SIZE, MU, ML, SMU);
    DlsMat b = NewBandMat(SIZE, MU, ML, SMU);
#endif
    sundials_ml_index p[SIZE] = { 0.0 };
    sunrealtype s[SIZE] = { 5.0, 15.0, 31.0, 53.0, 45.0 };

    BAND_ELEM(a,0,0) = 1.0;
    BAND_ELEM(a,0,1) = 2.0;

    BAND_ELEM(a,1,0) = 2.0;
    BAND_ELEM(a,1,1) = 2.0;
    BAND_ELEM(a,1,2) = 3.0;

    BAND_ELEM(a,2,1) = 3.0;
    BAND_ELEM(a,2,2) = 3.0;
    BAND_ELEM(a,2,3) = 4.0;

    BAND_ELEM(a,3,2) = 4.0;
    BAND_ELEM(a,3,3) = 4.0;
    BAND_ELEM(a,3,4) = 5.0;

    BAND_ELEM(a,4,3) = 5.0;
    BAND_ELEM(a,4,4) = 5.0;

    printf("initially: a=\n");
    print_mat(a);
    printf("\n");

#if SUNDIALS_LIB_VERSION >= 260
    {
	sunrealtype x[SIZE] = { 1.0,  2.0, 3.0, 4.0, 5.0 };
	sunrealtype y[SIZE] = { 0.0 };
	printf("matvec: y=\n");
#if 600 <= SUNDIALS_LIB_VERSION
	SUNDlsMat_BandMatvec(a, x, y);
#else
	BandMatvec(a, x, y);
#endif
	print_vec(y, SIZE);
	printf("\n");
    }
#endif

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_BandCopy(a, b, MU, ML);
    SUNDlsMat_BandScale(2.0, b);
#else
    BandCopy(a, b, MU, ML);
    BandScale(2.0, b);
#endif
    printf("scale copy x2: b=\n");
    print_mat(b);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_bandAddIdentity(b->cols, b->N, b->s_mu);
#else
    bandAddIdentity(b->cols, b->N, b->s_mu);
#endif
    printf("add identity: b=\n");
    print_mat(b);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_BandGBTRF(a, p);
#else
    BandGBTRF(a, p);
#endif
    printf("getrf: a=\n");
    print_factored_mat(a);
    printf("\n       p=\n");
    print_pivots(p, SIZE);
    printf("\n");

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_BandGBTRS(a, p, s);
#else
    BandGBTRS(a, p, s);
#endif
    printf("getrs: s=\n");
    print_vec(s, SIZE);

#if 600 <= SUNDIALS_LIB_VERSION
    SUNDlsMat_DestroyMat(a);
    SUNDlsMat_DestroyMat(b);
#else
    DestroyMat(a);
    DestroyMat(b);
#endif

    return 0;
}

